pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,329 @@
+  CHECK(options != NULL);
+
+  GoogleString controller_pid_filename(options->controller_pid_filename());
+  if (controller_pid_filename.empty()) {
+    handler_->Message(
+        kError, "Controller PID Filename should not be set to empty");
+    return "";
+  } else if (controller_pid_filename[0] == '/') {
+    return controller_pid_filename;
+  } else {
+    // Relative paths are interpreted relative to the directory in which the
+    // webserver would normally store the pid file for its master process.
+
+    GoogleString master_pid_filename = factory_->MasterPidFilename();
+    size_t final_slash = master_pid_filename.find_last_of('/');
+
+    return StrCat(
+        master_pid_filename.substr(0, final_slash), "/",
+        controller_pid_filename);
+  }
+}
+
+void ControllerManager::WriteControllerPid() {
+  GoogleString controller_pid_filename(ControllerPidFilename());
+  if (controller_pid_filename.empty()) {
+    return;
+  }
+
+  file_system_->WriteFile(controller_pid_filename.c_str(),
+                          Integer64ToString(getpid()),
+                          handler_);
+}
+
+pid_t ControllerManager::ControllerPid() {
+  GoogleString controller_pid_filename(ControllerPidFilename());
+  if (controller_pid_filename.empty()) {
+    return -1;
+  }
+
+  return ReadPidFromFile(controller_pid_filename);
+}
+
+void ControllerManager::KillControllerIfAlreadyRunning() {
+  pid_t controller_pid = ControllerPid();
+  if (controller_pid == -1) {
+    // Lookup failure, already logged.
+    return;
+  }
+
+  handler_->Message(
+      kInfo,
+      "About to send SIGKILL to babysitter process (%d) to tell it to "
+      "exit.  The controller will see it exit and exit as well.",
+      controller_pid);
+  // Babysitter has no cleanup to do, no reason to avoid SIGKILL.
+  int ret = kill(controller_pid, SIGKILL);
+  if (ret == -1) {
+    // Signal failed to send.
+    if (errno == ESRCH) {
+      // This is fine.  This is what usually happens if the controller exits
+      // abruptly and leaves behind a stale pid in its file.
+    } else if (errno == EPERM) {
+      handler_->Message(
+          kWarning,
+          "Permissions error trying to kill controller process at PID %d.",
+          controller_pid);
+    } else {
+      LOG(DFATAL) << "Got unexpected failure " << errno << " from kill.";
+    }
+    return;  // If the signal didn't go out, no point waiting for an effect.
+  }
+
+  for (int i = 0; i < 100; i++) {
+    // Wait for up to a second for it to die.
+    if (!IsProcessRunning(controller_pid)) {
+      return;
+    }
+    timer_->SleepUs(10 * 1000 /* 0.01s */);
+  }
+  handler_->Message(
+      kWarning, "Couldn't get controller process (%d) to exit", controller_pid);
+}
+
+void ControllerManager::RunController() {
+  handler_->Message(kInfo, "Controller running with PID %d", getpid());
+
+  // Would set up gRPC server here and pass control to its event loop.  Instead
+  // just hang out sleeping forever.
```

Comment:
The exit is handled by the watcher thread, right? May be worth mentioning that here.